overlord/configstate/configcore/watchdog.go

```diff
@@ -0,0 +1,142 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package configcore
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/snapcore/snapd/dirs"
+)
+
+var watchdogConfigKeys = map[string]bool{
+	"RuntimeWatchdogSec":  true,
+	"ShutdownWatchdogSec": true,
+}
+
+func init() {
+	// add supported configuration of this module
+	supportedConfigurations["core.watchdog.runtime-timeout"] = true
+	supportedConfigurations["core.watchdog.shutdown-timeout"] = true
+}
+
+func watchdogConfPath() string {
+	return filepath.Join(dirs.GlobalRootDir, "/etc/systemd/system.conf.d/")
+}
+
+func watchdogConfEnvironment() string {
+	return filepath.Join(watchdogConfPath(), "10-ubuntu-core-watchdog.conf")
+}
+
+func updateWatchdogConfig(config map[string]uint) error {
+	path := watchdogConfEnvironment()
+
+	configStr := []string{}
+	for k, v := range config {
+		if v > 0 {
+			configStr = append(configStr, fmt.Sprintf("%s=%d\n", k, v))
+		}
+	}
+	if len(configStr) == 0 {
+		// No config, remove file
+		if err := os.Remove(path); err != nil && !os.IsNotExist(err) {
+			return err
+		}
+		return nil
+	}
+
+	// We order the variables to have predictable output
+	sort.Strings(configStr)
+	content := "[Manager]\n" + strings.Join(configStr, "")
+
+	if err := os.MkdirAll(watchdogConfPath(), 0755); err != nil {
+		return err
+	}
+
+	return ioutil.WriteFile(path, []byte(content), 0644)
+}
+
+func handleWatchdogConfiguration(tr Conf) error {
+	config := map[string]uint{}
+
+	for _, key := range []string{"runtime-timeout", "shutdown-timeout"} {
+		output, err := coreCfg(tr, "watchdog."+key)
+		if err != nil {
+			return err
+		}
+		secs, err := getSystemdConfSeconds(output)
+		if err != nil {
+			return fmt.Errorf("cannot set timer to %q: %v", output, err)
+		}
+		switch key {
+		case "runtime-timeout":
+			config["RuntimeWatchdogSec"] = secs
+		case "shutdown-timeout":
+			config["ShutdownWatchdogSec"] = secs
+		}
+	}
+
+	if err := updateWatchdogConfig(config); err != nil {
```

Comment:
Ideally we'd not touch the configuration unless it has changed. We have a mechanism to tell what has changed in the configuration already.

Reply:
updateWatchdogConfig() changed to write the file only if the new content differs, as in other cases (and test added).

Reply:
Thanks for your update on this. I looked at the updated code and while its fine in itself I tweaked it a bit because we have some nice helpers for ensuring that a file is in a particular state (and does not get re-written etc). 

My suggestion is in https://github.com/snapcore/snapd/compare/master...mvo5:support-systemd-watchdog-tweaks?expand=1 (https://github.com/snapcore/snapd/commit/4cc89ab850d0655a2a18b46a57614dddf93d16ae). It also adds a test that the config is removed if all values are empty (iirc this was missing). 

I hope you like it, its slightly shorter and (I think) a bit more idiomatic in the snapd code-base context.

Reply:
We can of course also do this as a followup

Reply:
Thanks for the change, I have cherry-picked it and incorporated to the PR.